src/AI/AIUtils.java

```diff
@@ -170,11 +175,11 @@ public static GameAction moveTowardLocation(Unit unit, XYCoord destination, Game
 
     // Find the full path that would get this unit to the destination, regardless of how long. 
     Path path = Utils.findShortestPath(unit, destination, gameMap, true);
-    if( path.getPathLength() > 0 ) // Check that the destination is reachable at least in theory.
+    ArrayList<XYCoord> validMoves = Utils.findPossibleDestinations(unit, gameMap, false); // Find the valid moves we can make.
```

Comment:
Why not move this down to just above line 183 to where it's actually used? No reason to do this calculation if we don't need to. I know you check that `validMoves.size() > 0`, but that is trivially true if `path.getPathLength() > 0`.
I also don't like to pass in magic values; I'd declare a named temporary and pass that in just so it's obvious by reading what `false` means in this context.

Reply:
I'm not sure that it *is* trivially true. If I'm trapped in a sea of friendly units, I'll see a way out with infinite movement but won't actually be able to make any progress.

Reply:
All righty.